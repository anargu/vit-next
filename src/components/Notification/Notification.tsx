import React from 'react';
import { NotificationsProvider, showNotification as mantineNotification } from '@mantine/notifications';
import { DefaultMantineColor } from '@mantine/core';

export type NotificationProps = {
  children?: JSX.Element;
};

export type NotificationOptions = {
  color?: DefaultMantineColor | undefined | string,
}

export const showNotification = (message: string, title : string = "Information", options: NotificationOptions) => {
  mantineNotification({
    title: title,
    message: message,
    color: options?.color ?? "indigo",
  });
};

export const showDefaultNotification = (message: string, title : string = "Information") => {
  mantineNotification({
    message,
    title,
    color: "indigo",
  });
};


export const WithNotificationsProvider = (props : NotificationProps) : JSX.Element => {
  if (!props.children) return <div />;

  return <NotificationsProvider>{props.children}</NotificationsProvider>;
};
