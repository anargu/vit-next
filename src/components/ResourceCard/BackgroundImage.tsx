
export type BackgroundImageProps = {
  src : string,
  alt : string,
};

export const BackgroundImage = (props : BackgroundImageProps) => (
  <div
    role="img"
    alt={props.alt}
    className={`w-full h-full bg-contain bg-no-repeat`}
    style={{ backgroundImage: `url(${props.src})`}}
  />
);
