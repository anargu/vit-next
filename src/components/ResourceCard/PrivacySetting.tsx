
import { Switch } from '@mantine/core';
import React from 'react'

export type PrivacySettingProps = {
  isPublic : boolean,
  onChangePrivacy: (newValue : boolean) => void | Promise<void>,
};

export const PrivacySetting = (props : PrivacySettingProps) => {
  return (
    <div>
      <Switch
        label="Make it public"
        checked={props.isPublic}
        onChange={(event) => props.onChangePrivacy(event.currentTarget.checked)}
      />
    </div>
  )
};

