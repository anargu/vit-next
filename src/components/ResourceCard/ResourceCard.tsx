import { Resource, VITResource } from "../../core/entities";
import Image from "next/image"
import { useMemo } from "react";
import styled from 'styled-components';
import Link from "../../../public/assets/link.svg";
import Bookmark from "../../../public/assets/bookmark.svg";
import Share from "../../../public/assets/share.svg";

export type ResourceCardProps = {
  hit: VITResource,
};

export const ResourceCard = ({ hit } : ResourceCardProps) => {

  const resourceData = useMemo(() => Resource.fromVITResource(hit), [hit]);

  return (
    <div className="relative py-4 px-4 text-white h-full">
      <div className="absolute top-0 right-0 bottom-0 left-0 bg-black/60 z-20">
      </div>
      <div className="absolute top-0 right-0 bottom-0 left-0 z-10">
        <Image src={resourceData.imageSrc} alt={resourceData.imageAlt} layout="fill"/>
      </div>

      <div className="relative z-30 h-full grid grid-rows-[min-content_2fr_1fr_36px] gap-5">
        <div className="text-right"><span>{resourceData.timeAgo.toString()}</span></div>

        <p className={`py-0 max-h-[calc(2.5rem*4)]
          text-4xl text-multiline-overflow-mixin
          text-ellipsis [-webkit-line-clamp:4]`}>{resourceData.title}</p>

        <p className="self-center max-h-[calc(2rem*4)] text-2xl text-multiline-overflow-mixin text-ellipsis [-webkit-line-clamp:2]">{resourceData.description}</p>
        <ActionsStyled className="flex justify-around">
          <Link/>
          <Bookmark/>
          <Share/>
        </ActionsStyled>
      </div>
    </div>
  );
};

const ActionsStyled = styled.div`
  svg path {
    stroke: #FFF;
  }
`;
