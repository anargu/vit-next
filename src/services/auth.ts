import { signInWithRedirect, GoogleAuthProvider } from "firebase/auth";
import { doc, getDoc, setDoc, updateDoc } from "firebase/firestore";
import { auth, firestore, getRedirectResult } from "../firebase"

export type AuthenticatedUser = {
  id: string,
  email: string | null,
  displayName: string | null,
}

const googleProvider = new GoogleAuthProvider();

const signIn = async () => {
  await signInWithRedirect(auth, googleProvider)
};

const getSignInResult = async () => {
  return new Promise<AuthenticatedUser | null>((resolve, reject) => {
    getRedirectResult(auth)
      .then((result) => {
        const user = result?.user;

        if (!user) return resolve(null);

        resolve({
          id: user.uid,
          email: user.email,
          displayName: user.displayName,
        });
      }).catch((error) => {
        /* const errorCode = error.code; */
        /* const errorMessage = error.message; */
        /* // The email of the user's account used. */
        /* const email = error.customData.email; */
        console.error(`>>> Error: ${error.message}.\nCode: ${error.code}.\nUser:${error.customData?.email}`);

        reject(error);
      });
  });
};

const upsertUser = async (user : AuthenticatedUser) => {
  try {
    const docRef = doc(firestore, `users/${user.id}`);

    const userSnap = await getDoc(docRef);

    const upsertUserPromise = userSnap.exists()
      ? updateDoc(userSnap.ref, { lastSignIn: new Date() })
      : setDoc(userSnap.ref, {
          id: user.id,
          updatedAt: new Date(),
          createdAt: new Date(),
          lastSignIn: new Date(),
          email: user.email,
          displayName: user.displayName,
        });

    await upsertUserPromise;
  } catch (error) {
    console.error(`Error at upsertUser: ${error}`);
  }
};

export {
  signIn,
  getSignInResult,
  upsertUser,
};

