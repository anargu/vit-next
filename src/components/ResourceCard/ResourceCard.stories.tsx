import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import { ResourceCard, ResourceCardProps } from "./ResourceCard";

import { faker } from "@faker-js/faker";
import { VITResource } from '@/src/core/entities';
// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'UI/ResourceCard',
  component: ResourceCard,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
  },
} as ComponentMeta<typeof ResourceCard>;

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof ResourceCard> = (args : ResourceCardProps) => (
  <div className="bg-gray-100 w-[300px] h-[400px]"><ResourceCard {...args} /></div>
);

const mockHitData = () : VITResource => ({
  id: faker.datatype.uuid(),
  og_image: "https://source.unsplash.com/random/50x50",
  keyphrase: "man holding a beer",
  date_created: faker.datatype.datetime().toISOString(),
  og_title: "Inter Link",
  og_description: "This is a link, interlinked",
  url_title: "The link",
  url: "https://www.google.com",
});

export const Default = Template.bind({});
// More on args: https://storybook.js.org/docs/react/writing-stories/args
Default.args = {
  hit: mockHitData(),
};

export const LongText = Template.bind({});

LongText.args = {
  hit: {
    ...mockHitData(),
    og_title: faker.lorem.paragraph(2),
    og_description: faker.lorem.paragraphs(4),
  }
};

const FluidTemplate: ComponentStory<typeof ResourceCard> = (args : ResourceCardProps) => (
  <div className="bg-gray-100 h-[400px]"><ResourceCard {...args} /></div>
);

export const FluidLayout = FluidTemplate.bind({});

FluidLayout.args = {
  hit: {
    ...mockHitData(),
    og_title: faker.lorem.paragraph(2),
    og_description: faker.lorem.paragraphs(4),
  }
};

export const NoImage = Template.bind({});
// More on args: https://storybook.js.org/docs/react/writing-stories/args
NoImage.args = {
  hit: {
    ...mockHitData(),
    og_image: null,
  }
};
