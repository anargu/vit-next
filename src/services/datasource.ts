import { APIResponse } from "@/pages/api/entities";
import { MetascrapperResponse } from "@/pages/api/metadata";
import { firestore } from "../firebase";
import { addDoc, updateDoc, collection, doc, DocumentData, getDoc, onSnapshot, query, QuerySnapshot, where, orderBy, setDoc, DocumentSnapshot } from "firebase/firestore";
import { SAVED_LINK_KEY } from "../components/ResourceCard/ResourceCard";
import { AuthenticatedUser, LinkPrivacy, Resource, User, VITResource } from "../core/entities";
import { DataTransformer, Link } from "./entities";

const fetchUser = async (id : string) : Promise<User> => {
  const userRef = doc(firestore, `users/${id}`);
  const userSnap = await getDoc(userRef);

  const user = userSnap.data();

  return user as User;
};

const listenFeedLinks = (
  onSnapshotCallback : (snapshot : QuerySnapshot<DocumentData>) => void
) => {
  const linksRef = collection(firestore, `feed`);

  const q = query(linksRef, where("deleted", "==", false), orderBy("createdAt", "desc"))

  const unSub = onSnapshot(q, onSnapshotCallback,
    (err) => {
      console.error(err);
    });
  
  return unSub;
};

const listenLinkByID = (
  id : string,
  onSnapshotCallback : (resource : Resource) => void
) => {
  const linkRef = doc(firestore, `links/${id}`);

  const unSub = onSnapshot(
    linkRef,
    (linkSnapshot) => onSnapshotCallback(DataTransformer.fromLinktoResource({ ...linkSnapshot.data() as Link, id: linkSnapshot.id, })),
    (err) => {
      console.error(err);
    });
  
  return unSub;
};

const listenLinksFromUser = (
  id : string,
  onSnapshotCallback : (snapshot : QuerySnapshot<DocumentData>) => void
) => {
  const linksRef = collection(firestore, `links`);

  const q = query(linksRef, where("byId", "==", id), where("deleted", "==", false), orderBy("createdAt", "desc"))

  const unSub = onSnapshot(q, onSnapshotCallback,
    (err) => {
      console.error(err);
    });
  
  return unSub;
};

const insertLink = async (data : Resource, by : User) => {
  const linksRef = collection(firestore, `links`);

  let modifiedData : any = { ...data };
  delete modifiedData?.id;

  // TODO: Transform datetime fields
  await addDoc(linksRef, {
    ...modifiedData,
    deleted: false,
    byId: by.id,
    by: {
      id: by.id,
      email: by.email,
    },
  });
};

const updateLinkPrivacy = async (resourceId : string, privacy : LinkPrivacy) => {
  const feedDocRef = doc(firestore, `feed/${resourceId}`);
  const linkRef = doc(firestore, `links/${resourceId}`);

  await updateDoc(linkRef, {
    updatedAt: new Date(),
    isPublic: (privacy as LinkPrivacy) === LinkPrivacy.PUBLIC
      ? true
      : false,
  });

  if (privacy === LinkPrivacy.PRIVATE) {

    await updateDoc(feedDocRef, {
      deleted: true,
      updatedAt: new Date(),
    });

    return;
  }

  const linkSnapshot = await getDoc(linkRef);

  await setDoc(feedDocRef, {
    ...linkSnapshot.data(),
    id: linkSnapshot.id,
    updatedAt: new Date(),
    deleted: false,
  }, { merge: true });
};

const fetchMetadataFromURL = async (url : string) : Promise<Resource> => {

  const response = await fetch(`/api/metadata?url=${url}`);
  const resourceMetadata = await response.json() as APIResponse;

  if (!resourceMetadata.isOk) return new Resource({
    url: url,
    url_title: url,
    date_created: new Date().toISOString(),
    og_description: "",
  } as VITResource)

  return new Resource({
    date_created: new Date().toISOString(),
    url: (resourceMetadata.data as MetascrapperResponse)?.url,
    url_title: (resourceMetadata.data as MetascrapperResponse)?.title,
    og_image: (resourceMetadata.data as MetascrapperResponse)?.image,
    keyphrase: (resourceMetadata.data as MetascrapperResponse)?.image,
    og_title: (resourceMetadata.data as MetascrapperResponse)?.title,
    og_description: (resourceMetadata.data as MetascrapperResponse)?.description,
  } as VITResource)
};

const migrateLocalData = async (authenticatedUser : AuthenticatedUser) => {
  try {
    if (!authenticatedUser.id) return null;

    const rawData = localStorage.getItem(SAVED_LINK_KEY) ?? "[]";
    const localSavedPosts : VITResource[] = JSON.parse(rawData);

    if (!localSavedPosts || localSavedPosts.length === 0) return null;

    const linksRef = collection(firestore, `links`);

    // Transform
    const insertLocalLinkPromises = localSavedPosts.map(Resource.fromVITResource).map((localPost) => {
      const localPost_ : any = localPost;

      delete localPost_.id;

      return addDoc(linksRef, {
        ...localPost_,
        deleted: false,
        byId: authenticatedUser.id,
        by: { ...authenticatedUser },
      });
    });

    // Upload data
    await Promise.all(insertLocalLinkPromises);

    // Clean local data.
    localStorage.removeItem(SAVED_LINK_KEY);

    return true;
  } catch (error) {
    console.error(error);
    throw new Error("Local data migration failed. Error: ${error?.message}");
  }
};

const unsaveLink = async (linkId : string) => {
  const linksRef = doc(firestore, `links/${linkId}`);

  await updateDoc(linksRef, {
    deleted: true,
  });
};

export {
  fetchUser,
  listenFeedLinks,
  listenLinksFromUser,
  listenLinkByID,
  insertLink,
  unsaveLink,
  fetchMetadataFromURL,
  updateLinkPrivacy,
  migrateLocalData,
};
