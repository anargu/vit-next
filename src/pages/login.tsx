import { useRouter } from "next/router";
import { useEffect, useState } from "react";
import { CircularLoader } from "../components/Loaders";
import { showDefaultNotification } from "../components/Notification/Notification";
import { getSignInResult, signIn, upsertUser } from "../services/auth";
import { migrateLocalData } from "../services/datasource";

export const LoginPage = () => {
  const router = useRouter();

  const [hasError, setHasError] = useState<any>(null);
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    const fetchSignInResult = async () => {
      try {
        setIsLoading(true);
        const authenticatedUser = await getSignInResult();

        if (!authenticatedUser) return setIsLoading(false);

        // Creates user
        await upsertUser(authenticatedUser);

        // Migrate user data if any
        if (await migrateLocalData(authenticatedUser)) {
          showDefaultNotification(
            "Migration of local data succeeded!",
            "Info",
          );
        }

        setIsLoading(false);

        return await router.push("/");
      } catch (error) {
        console.error(`>>> error ${error}`);
        setHasError(error);
      }

      setIsLoading(false);
    };

    fetchSignInResult();
  }, []);

  return <div className="h-screen flex flex-col justify-center items-center">
    <button
      onClick={(e) => {
        e.preventDefault();

        signIn();
      }}
      type="button"
      className="text-white bg-[#4285F4] hover:bg-[#4285F4]/90 focus:ring-4 focus:outline-none focus:ring-[#4285F4]/50 font-medium rounded-lg text-sm px-5 py-2.5 text-center inline-flex items-center dark:focus:ring-[#4285F4]/55 mr-2 mb-2"
    >
      <svg className="mr-2 -ml-1 w-4 h-4" aria-hidden="true" focusable="false" data-prefix="fab" data-icon="google" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 488 512"><path fill="currentColor" d="M488 261.8C488 403.3 391.1 504 248 504 110.8 504 0 393.2 0 256S110.8 8 248 8c66.8 0 123 24.5 166.3 64.9l-67.5 64.9C258.5 52.6 94.3 116.6 94.3 256c0 86.5 69.1 156.6 153.7 156.6 98.2 0 135-70.4 140.8-106.9H248v-85.3h236.1c2.3 12.7 3.9 24.9 3.9 41.4z"></path></svg>
      Sign in with Google
    </button>

    {hasError
    ? (<div className="p-4 mb-4 text-sm text-red-700 bg-red-100 rounded-lg dark:bg-red-200 dark:text-red-800" role="alert">
          <span className="font-medium">Oops! there is an Error:</span> {hasError?.message ?? "Please try again"}
      </div>)
    : null}

    {isLoading
    ? (<div className="p-4 mb-4 text-sm text-stone-800 bg-amber-300 rounded-lg dark:bg-red-200 dark:text-red-800" role="alert">
          <span className="font-regular">We are syncing your data. Please wait a moment.</span>
          <span className="px-2"><CircularLoader size="18px" colorClassname="text-stone-800" /></span>
      </div>)
    : null}

  </div>;
}

