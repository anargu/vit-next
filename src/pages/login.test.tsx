import { render, waitFor } from "@testing-library/react";
import { LoginPage } from "./login";
import { getSignInResult, upsertUser } from "../services/auth";
import { useRouter } from "next/router";
import { SAVED_LINK_KEY } from "../components/ResourceCard/ResourceCard";
import { migrateLocalData } from "../services/datasource";
import { WithNotificationsProvider } from "../components/Notification/Notification";

console.error = jest.fn();

jest.mock('../services/auth', () => {
  return {
    getSignInResult: jest.fn(),
    upsertUser: jest.fn(),
  };
});

jest.mock('next/router', () => {
  return {
    useRouter: jest.fn(),
  }
});

describe("Login", () => {
  it("renders login page", () => {
    const { container } = render(<LoginPage />);

    expect(container).toMatchSnapshot();
  });

  it("shows a loading visual feedback when it is loading for result while login", async () => {
    const loadingMessage = "We are syncing your data. Please wait a moment.";

    (getSignInResult as jest.Mock).mockImplementation(async () => {
      await new Promise((r) => setTimeout(r, 2000));

      return Promise.resolve({
        id: "1",
        displayName: "John Doe",
        email: "john@gmail.com",
      });
    });

    (upsertUser as jest.Mock).mockReturnValue(() => Promise.resolve(null));

    const pushMock = jest.fn();

    (useRouter as jest.Mock).mockImplementation(() => ({
      push: pushMock,
    }))

    const { getByText } = render(<LoginPage />);

    expect(getByText(loadingMessage)).toBeTruthy();
  });

  it("returns to home page when it returns authenticated user after login ", async () => {
    localStorage.removeItem(SAVED_LINK_KEY);

    (getSignInResult as jest.Mock).mockImplementation(() => {
      return Promise.resolve({
        id: "1",
        displayName: "John Doe",
        email: "john@gmail.com",
      });
    });

    (upsertUser as jest.Mock).mockReturnValue(() => Promise.resolve(null));

    const pushMock = jest.fn();

    (useRouter as jest.Mock).mockImplementation(() => ({
      push: pushMock,
    }))

    render(<LoginPage />);

    await waitFor(() => {
      expect(getSignInResult as jest.Mock).toHaveBeenCalled();
      expect(upsertUser as jest.Mock).toHaveBeenCalled();
      expect(pushMock as jest.Mock).toHaveBeenCalled();
    });
  });

  // Error handling
  it("shows error message alert when login failed", async () => {
    const errorMessage = "Failed authentication";

    (getSignInResult as jest.Mock).mockImplementation(() => {
      return Promise.reject(new Error(errorMessage));
    });

    const { getByText } = render(<LoginPage />);

    await waitFor(() => {
      getByText(errorMessage);
    });
  });

  it("shows general error message alert when login failed and not error message is provided", async () => {
    const errorMessage = "Please try again";

    (getSignInResult as jest.Mock).mockReturnValue(Promise.reject({}));

    const { getByText } = render(<LoginPage />);

    await waitFor(() => {
      getByText(errorMessage);
    });
  });

  describe("User Data Migration", () => {
    it("throws a message if upload of transformed data fails", async () => {
      (getSignInResult as jest.Mock).mockImplementation(() => {
        return Promise.resolve({
          id: "1",
          displayName: "John Doe",
          email: "john@gmail.com",
        });
      });

      (migrateLocalData as jest.Mock).mockImplementation(() => Promise.reject(new Error("it failed")));

      const { getByText } = render(<LoginPage />);

      await waitFor(() => {
        getByText(("it failed"));
      });
    });

    it("shows notification if data migration succeeds", async () => {
      const message = "Migration of local data succeeded!";

      (migrateLocalData as jest.Mock).mockImplementation(() => Promise.resolve(true));

      const { getByText } = render(
        <WithNotificationsProvider>
          <LoginPage />
        </WithNotificationsProvider>
      );

      await waitFor(() => {
        getByText((message));
      });
    });
  });
});

