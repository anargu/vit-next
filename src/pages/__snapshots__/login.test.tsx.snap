// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders login page 1`] = `
<div>
  <div
    class="h-screen flex flex-col justify-center items-center"
  >
    <button
      class="text-white bg-[#4285F4] hover:bg-[#4285F4]/90 focus:ring-4 focus:outline-none focus:ring-[#4285F4]/50 font-medium rounded-lg text-sm px-5 py-2.5 text-center inline-flex items-center dark:focus:ring-[#4285F4]/55 mr-2 mb-2"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="mr-2 -ml-1 w-4 h-4"
        data-icon="google"
        data-prefix="fab"
        focusable="false"
        role="img"
        viewBox="0 0 488 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M488 261.8C488 403.3 391.1 504 248 504 110.8 504 0 393.2 0 256S110.8 8 248 8c66.8 0 123 24.5 166.3 64.9l-67.5 64.9C258.5 52.6 94.3 116.6 94.3 256c0 86.5 69.1 156.6 153.7 156.6 98.2 0 135-70.4 140.8-106.9H248v-85.3h236.1c2.3 12.7 3.9 24.9 3.9 41.4z"
          fill="currentColor"
        />
      </svg>
      Sign in with Google
    </button>
    <div
      class="p-4 mb-4 text-sm text-stone-800 bg-amber-300 rounded-lg dark:bg-red-200 dark:text-red-800"
      role="alert"
    >
      <span
        class="font-regular"
      >
        We are syncing your data. Please wait a moment.
      </span>
      <span
        class="px-2"
      >
        <svg
          aria-hidden="true"
          class="animate-spin inline-block text-stone-800"
          data-icon="loading"
          fill="currentColor"
          focusable="false"
          height="18px"
          viewBox="0 0 1024 1024"
          width="18px"
        >
          <path
            d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;
